job_creator.php

```diff
@@ -0,0 +1,270 @@
+                    'js' => true
+                ]);
+            }
+        }
+
+        // extra jobs
+        foreach ($extraJobs as $arr) {
+            $matrix['include'][] = $this->createJob(0, $arr);
+        }
+
+        // convert everything to strings and sanatise values
+        foreach ($matrix['include'] as $i => $job) {
+            foreach ($job as $key => $val) {
+                if ($val === true) {
+                    $val = 'true';
+                }
+                if ($val === false) {
+                    $val = 'false';
+                }
+                // all values must be strings
+                $val = (string) $val;
+                // remove any dodgy characters
+                $val = str_replace(["\r", "\n", "\t", "'", '"', '&', '|'], '', $val);
+                // ascii chars only - https://www.regular-expressions.info/posixbrackets.html
+                $val = preg_replace('#[^\x20-\x7E]#', '', $val);
+                // limit name_suffix length and be strict as it is used in the artifact file name
+                if ($key === 'name_suffix') {
+                    if (strlen($val) > 20) {
+                        $val = preg_replace('#[^a-zA-Z0-9_\- ]#', '', $val);
+                        $val = substr($val, 0, 20);
+                    }
+                }
+                // be strict with composer_require_extra
+                if ($key === 'composer_require_extra') {
+                    $val = preg_replace('#[^A-Za-z0-9\-\.\^\/~: ]#', '', $val);
+                }
+                // add value back to matrix
+                $matrix['include'][$i][$key] = $val;
+            }
+        }
+
+        // job/artifacts names
+        foreach ($matrix['include'] as $i => $job) {
+            $name = [
+                $job['php']
+            ];
+            if ($job['composer_args'] == '--prefer-lowest') {
+                $name[] = 'prf-low';
+            }
+            $name[] = $job['db'];
+            if ($job['phpunit'] == 'true') {
+                $name[] = 'phpunit';
+                $name[] = $job['phpunit_suite'];
+            }
+            if ($job['endtoend'] == 'true') {
+                $name[] = 'endtoend';
+                $name[] = $job['endtoend_suite'];
+            }
+            if ($job['js'] == 'true') {
+                $name[] = 'js';
+            }
+            if ($job['phpcoverage'] == 'true') {
+                $name[] = 'phpcoverage';
+            }
+            if ($job['phplinting'] == 'true') {
+                $name[] = 'phplinting';
+            }
+            $name[] = $job['name_suffix'];
+            $name = array_filter($name);
+            $matrix['include'][$i]['name'] = implode(' ', $name);
+        }
+
+        // output json, keep it on a single line so do not use pretty print
+        $json = json_encode($matrix);
+        $json = preg_replace("#\n +#", "\n", $json);
+        $json = str_replace('\\/', '/', $json);
+        $json = str_replace("\n", '', $json);
```

Comment:
In what situation will there be a newline character in the JSON?

Reply:
default output of json_encode I think

Reply:
Okay. No harm in keeping it here, I was mostly just curious.